Improve optimizer to recognize more affine handlers

dolio opened the pull request:
This refactors things a bit to be able to apply extra optimizations before trying to recognize affine handlers. The extra optimizations are only safe in an affine context, because they may rearrange values on the stack, or inline expressions in ways that won't be properly reconstructible on serialization. However, these differences don't matter, because stacks are not captured in affine contexts.

The improvements allow recognizing affine handlers that are significantly more complicated than the most obvious examples. For instance, previously, if a 'nice' affine handler were written such that the outermost definition did anything besides immediately handling the forced thunk, it basically wouldn't work. That is no longer a restriction.

This allows some complex handlers in our libraries to be recognized correctly.

I've added some test cases, too that show things that previously didn't work (I think). In particular, some kind of odd recursive handler patterns now work (although the example here is not as elaborate as some of our actual handlers).

dolio committed:
Improve optimizer to recognize more affine handlers

This refactors things a bit to be able to apply extra optimizations
before trying to recognize affine handlers. The extra optimizations
are only safe in an affine context, because they may rearrange values
on the stack, or inline expressions in ways that won't be properly
reconstructible on serialization. However, these differences don't
matter, because stacks are not captured in affine contexts.

The improvements allow recognizing affine handlers that are
significantly more complicated than the most obvious examples. For
instance, previously, if a 'nice' affine handler were written such
that the outermost definition did anything besides immediately
handling the forced thunk, it basically wouldn't work. That is no
longer a restriction.

dolio committed:
Further expand affine recognition

Allows even more cases of the `HandledThunk` peephole optimization in
affine contexts.

Allows `replaceLinearBody` to apply to include some bindings prefixing
the inlined tail.

dolio committed:
Add some more affine handler test cases

dolio committed:
Add a bit more complexity to the test case.

dolio committed:
automatically run ormolu

ceedubs reviewed (approved):
I can't help in reviewing the code, but I checked that the Unison Cloud integration tests pass with this change.

pchiusano reviewed (approved):
Nice!

dolio committed:
Some assorted cleanup of optimizer

- Switch affinePostOptimize to just use the normal optimizer,
  incorporating the optimizations into the main peephole function
- Fix the call to affinePostOptimize, which wasn't actually
  running on the newly generated affine handler code
- Remove some unused arguments from some functions
- Add comments explaining the peephole optimizations more

dolio commented:
Not sure what's going on with the windows build. The racket stuff can't find some module?

dolio commented:
BTW, I ran base, cloud, json, codec and orderator tests, and they passed (except for the one broken test in one of those packages).

There might be a problem with some cloud I/O tests. I'm looking into that.

aryairani committed:
Merge branch 'trunk' into topic/affine-improvements

aryairani commented:
Erk sorry I went on a merging spree and missed the thing about the cloud I/O tests this morning. We can revert it if needed.

ceedubs commented:
All seems to be good with the Unison Cloud integration tests :+1:

aryairani commented:
Phew
